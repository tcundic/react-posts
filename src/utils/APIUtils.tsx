import {apiEndpoint} from "../config";
import RequestOptions from "../interfaces/Request";
import Post from "../interfaces/Post";

const request = (options: RequestOptions, additionalHeaders?: Object): Promise<any> => {
    const headers = {
        'Content-Type': 'application/json',
        ...additionalHeaders
    };

    const defaults = {headers};
    options = Object.assign({}, defaults, options, {crossDomain: true});

    return fetch(options.url, options)
        .then(response => response.json());
};

export function getPosts(): Promise<Post[]> {
    return request({
        url: `${apiEndpoint}posts`,
        method: 'GET'
    });
}

export function getPostById(id: number): Promise<Post> {
    return request({
        url: `${apiEndpoint}posts/${id}`,
        method: 'GET'
    });
}
