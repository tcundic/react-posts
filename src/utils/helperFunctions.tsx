export const greet = (greeting: string, componentName: string): void => console.log(`${greeting} ${componentName}`);