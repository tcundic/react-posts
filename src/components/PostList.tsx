import Post from "../interfaces/Post";
import React, {useEffect} from "react";
import {greet} from "../utils/helperFunctions";
import {NavLink} from "react-router-dom";

const PostList = ({greeting, posts}: {greeting: string, posts: Post[]}) => {

    useEffect(() => greet(greeting, PostList.name), [greeting]);

    return (
        <div className="posts-list mt-6">
            {posts.map(post => (
                <div key={post.id}>
                    <NavLink className="navbar-item" to={`/post/${post.id}`}>
                        <h2>{post.title}</h2>
                    </NavLink>
                    <div className="post-username-caption padding-x">@{post.user}</div>
                    <h3 className="padding-x mt-4">Comments</h3>
                    {post.comments.map(comment => (
                        <div className="post-comment padding-x" key={`${post.id}-${comment.id}`}>
                            {comment.content}
                        </div>
                    ))}
                </div>
            ))}
        </div>
    );
};

export default PostList;