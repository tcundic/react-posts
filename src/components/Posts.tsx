import {useEffect, useState} from "react";
import {greet} from "../utils/helperFunctions";
import {getPosts} from "../utils/APIUtils";
import Post from "../interfaces/Post";
import PostList from "./PostList";
import useFilter from "./useFilter";

const Posts = ({greeting}: {greeting: string}) => {

    const [posts, setPosts] = useState<Post[]>([]);
    const [filteredPosts, setFilteredPosts] = useState<Post[]>(posts);
    const [username, usernameInput] = useFilter({greeting, type: "text", placeholder: "Enter username"});

    useEffect(() => {
        greet(greeting, Posts.name);
    }, [greeting]);

    useEffect(() => {
        getPosts().then((posts: Post[]) => {
            setPosts(posts);
            setFilteredPosts(posts);
        });
    }, []);

    useEffect(() => {
        // @ts-ignore
        setFilteredPosts(posts.filter((post: Post) => post.user.includes(username)));
    }, [username, posts]);

    return (
        <div className="mt-6 mx-6">
            <div className="posts-header">
                <h1>Posts</h1>
                {usernameInput}
            </div>
            <PostList posts={filteredPosts} greeting={greeting} />
        </div>
    );
};

export default Posts;