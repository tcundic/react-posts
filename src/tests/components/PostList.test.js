import Posts from "../../components/Posts";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as Greet from "../../utils/helperFunctions";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import PostList from "../../components/PostList";
import posts from "../fixtures/posts";

const greeting = "Hello from";
const name = PostList.name;

test('should render PostList correctly', () => {
    const wrapper = shallow(<PostList greeting={greeting} posts={posts} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('NavLink')).toHaveLength(2);
});

test('should print PostList greeting', () => {
    const greetSpy = jest.spyOn(Greet, 'greet');
    mount(<BrowserRouter><PostList greeting={greeting} posts={posts}/></BrowserRouter>);
    expect(greetSpy).toHaveBeenLastCalledWith(greeting, name);
});
