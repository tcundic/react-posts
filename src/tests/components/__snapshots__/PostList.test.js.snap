// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PostList correctly 1`] = `
<div
  className="posts-list mt-6"
>
  <div
    key="1"
  >
    <NavLink
      className="navbar-item"
      to="/post/1"
    >
      <h2>
        Post 1
      </h2>
    </NavLink>
    <div
      className="post-username-caption padding-x"
    >
      @
      user 1
    </div>
    <h3
      className="padding-x mt-4"
    >
      Comments
    </h3>
    <div
      className="post-comment padding-x"
      key="1-1"
    >
      comment 1
    </div>
    <div
      className="post-comment padding-x"
      key="1-2"
    >
      comment 2
    </div>
  </div>
  <div
    key="2"
  >
    <NavLink
      className="navbar-item"
      to="/post/2"
    >
      <h2>
        Post 2
      </h2>
    </NavLink>
    <div
      className="post-username-caption padding-x"
    >
      @
      user 2
    </div>
    <h3
      className="padding-x mt-4"
    >
      Comments
    </h3>
  </div>
</div>
`;
